src/ar100/main.c

```diff
@@ -0,0 +1,176 @@
+// Main entry point for ar100
+//
+// Copyright (C) 2020-2021  Elias Bakken <elias@iagent.no>
+//
+// This file may be distributed under the terms of the GNU GPLv3 license.
+
+#include <stdint.h>     // uint32_t
+#include <string.h>
+#include "board/misc.h" // dynmem_start
+#include "board/irq.h"  // irq_disable
+#include "command.h"    // shutdown
+#include "generic/timer_irq.h"  // timer_dispatch_many
+#include "sched.h"      // sched_main
+
+#include "asm/spr.h"
+#include "util.h"
+#include "gpio.h"
+#include "serial.h"
+#include "prcm.h"
+#include "timer.h"
+
+DECL_CONSTANT_STR("MCU", "ar100");
+
+#define RESET_VECTOR 0x0100
+
+static struct task_wake console_wake;
+static uint8_t receive_buf[1024];
+static int receive_pos;
+static char dynmem_pool[2 * 1024];
+
+void *
+dynmem_start(void)
+{
+    return dynmem_pool;
+}
+
+void *
+dynmem_end(void)
+{
+    return &dynmem_pool[sizeof(dynmem_pool)];
+}
+
+void
+irq_disable(void)
+{
+}
+
+void
+irq_enable(void)
+{
+}
+
+irqstatus_t
+irq_save(void)
+{
+    return 0;
+}
+
+void
+irq_restore(irqstatus_t flag)
+{
+}
+
+void
+irq_wait(void)
+{
+    irq_poll();
+}
+
+void
+irq_poll(void)
+{
+    if(timer_interrupt_pending()) {
+        timer_clear_interrupt();
+        uint32_t next = timer_dispatch_many();
+        timer_set(next);
+    }
+    if(r_uart_fifo_rcv())
+        sched_wake_task(&console_wake);
+}
+
+/****************************************************************
+* Console IO
+****************************************************************/
+
+// Process any incoming commands
+void
+console_task(void)
+{
+    if (!sched_check_wake(&console_wake))
+        return;
+
+    int ret = 0;
+    for(int i=0; i<r_uart_fifo_rcv(); i++) {
+        receive_buf[receive_pos + ret++] = r_uart_getc();
+    }
+    if(!ret)
+        return;
+
+    int len = receive_pos + ret;
+    uint_fast8_t pop_count, msglen = len > MESSAGE_MAX ? MESSAGE_MAX : len;
+    ret = command_find_and_dispatch(receive_buf, msglen, &pop_count);
+    if (ret) {
+        len -= pop_count;
+        if (len) {
+            memcpy(receive_buf, &receive_buf[pop_count], len);
+            sched_wake_task(&console_wake);
+        }
+    }
+    receive_pos = len;
+}
+DECL_TASK(console_task);
```

Comment:
Can this uart code utilize the helpers in generic/serial_irq.c ?

Reply:
Yes, I think some of this code can be removed in favor of what is in generic/serial_irq.c. Regarding your question about interrupts, there is support for UART interrupt, which would match better with the existing code base. It is a bit of work though, so if it's possible to do it in a separate PR, that would be great.